planner/optimize.go

```diff
@@ -295,7 +295,21 @@ func optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		}
 		if preparedPointer, ok := sctx.GetSessionVars().PreparedStmts[execID]; ok {
 			if preparedObj, ok := preparedPointer.(*core.CachedPrepareStmt); ok && preparedObj.ForUpdateRead {
+				// Actually, `IS` is read only for all the travel path around optimize and exec. The `IS` and `TxnCtn.IS`
+				// should always be the same, otherwise, it will cause some check fails. For example. `checkUpdateList`
+				// panics in the plan building phase.
+				//
+				//  ...
+				//  optimize(is) -> OptimizePreparedPlan(is) -> ... `is` changed
+				//  runstmt(is)
+				//  ...
+				// The modification of `is` is in the deep call chain. once func the backtracking to the outer, since
+				// the `is` struct value copy model, the outer `is` is still the old one before calling the runstmt.
+				//
+				// while the session TxnCtx hold the new `IS`, we can substitute the outer `is` once we found the schema
+				// version difference before calling the runstmt
 				is = domain.GetDomain(sctx).InfoSchema()
+				sctx.GetSessionVars().TxnCtx.InfoSchema = is
```

Comment:
` I found 2pc relied on this behavior.` I don't find this? where is it?

Reply:
I only find it works on the `resolveAccessPaths`

Reply:
you are right. these two information can be different! may be for this sort of `forUpdateRead` case, they should be the same.

Reply:
2pc relied on schema checker and amender. Changing infoschema may bypass the checker and fail the correctness of amender.

You reminds me that I've suggested to remove these lines, because I have moved the logic of choosing correct infoschema to the outside https://github.com/pingcap/tidb/pull/26759/files#r679868151: these lines should be stub, the correct infoschema has been chosen. But got refused for risks, because the transaction reviewer seems can not track the whole execution path, too. 

I've looked more closely, `server/conn` will dispatch `prepare` and `execute` into the one in session, so yes, the correct infoschema will be passed down from session down to here. And chasing a even newer infoschema between two closely coupled `optimize` and `execute` is non-sense for me.

Reply:
We need a whole new proposal to define the behavior of concurrent DDL. Sync the latest = use the global latest, so all sessions will share the same infoschema as long as possible. And a talk to the transaction and runtime team is must.